File: backend-nodejs/src/core/game-resolver.ts
/**
 * Game name resolver for Steam app IDs.
 * 
 * This module handles resolving Steam app IDs to human-readable game names using multiple fallback methods.
 */

import { readFile, writeFile, mkdir } from 'fs/promises';
import { existsSync } from 'fs';
import { join, dirname } from 'path';
import { homedir } from 'os';
import type { AppManifest, LibraryCache } from '../types/index.js';

export class GameNameResolver {
  private readonly steamUserId: string;
  private readonly steamPath: string;
  private readonly steamappsPath: string;
  private readonly libraryCachePath: string;
  private readonly shortcutsPath: string;
  private readonly cacheFile: string;
  private readonly cachedMappings: Map<string, string> = new Map();

  // Manual mappings for games not found in other sources
  private readonly manualMappings: Record<string, string> = {
    '3241660': 'R.E.P.O.',
    // Add more mappings as needed
  };

  // System apps to filter out (optional)
  private readonly systemApps: Set<string> = new Set([
    'steam linux runtime',
    'proton',
    'steamworks',
    'steam client'
  ]);

  /**
   * Initialize the Game Name Resolver.
   * 
   * @param steamUserId - Steam user ID (default: "107256425")
   */
  constructor(steamUserId: string = '107256425') {
    this.steamUserId = steamUserId;
    this.steamPath = join(homedir(), '.local', 'share', 'Steam');
    this.steamappsPath = join(this.steamPath, 'steamapps');
    this.libraryCachePath = join(this.steamPath, 'userdata', steamUserId, 'config', 'librarycache');
    this.shortcutsPath = join(this.steamPath, 'userdata', steamUserId, 'config', 'shortcuts.vdf');
    this.cacheFile = join('instance', 'data', 'app_name_cache.json');

    // Load cached mappings
    this.loadCache();
  }

  /**
   * Load cached app ID to name mappings.
   * 
   * @returns Promise that resolves when cache is loaded
   */
  private async loadCache(): Promise<void> {
    if (existsSync(this.cacheFile)) {
      try {
        const content = await readFile(this.cacheFile, 'utf-8');
        const data = JSON.parse(content) as Record<string, string>;
        
        // Convert to Map
        for (const [appId, name] of Object.entries(data)) {
          this.cachedMappings.set(appId, name);
        }
      } catch (error) {
        console.warn(`Warning: Could not load app name cache: ${error}`);
      }
    }
  }

  /**
   * Save app ID to name mappings to cache.
   * 
   * @returns Promise that resolves when cache is saved
   */
  private async saveCache(): Promise<void> {
    try {
      await mkdir(dirname(this.cacheFile), { recursive: true });
      
      // Convert Map to object
      const data: Record<string, string> = {};
      for (const [appId, name] of this.cachedMappings) {
        data[appId] = name;
      }
      
      await writeFile(this.cacheFile, JSON.stringify(data, null, 2), 'utf-8');
    } catch (error) {
      console.warn(`Warning: Could not save app name cache: ${error}`);
    }
  }

  /**
   * Get game name from app manifest file (primary method).
   * 
   * @param appId - Steam app ID
   * @returns Game name or null if not found
   */
  private async getNameFromAppManifest(appId: string): Promise<string | null> {
    const manifestPath = join(this.steamappsPath, `appmanifest_${appId}.acf`);
    
    if (!existsSync(manifestPath)) {
      return null;
    }

    try {
      const content = await readFile(manifestPath, 'utf-8');
      
      // Simple parsing for the name field
      for (const line of content.split('\n')) {
        if (line.includes('"name"') && line.includes('\t')) {
          // Extract name from line like: "name"		"Game Name"
          const parts = line.split('\t');
          if (parts.length >= 3) {
            const name = parts[parts.length - 1].trim().replace(/^"/, '').replace(/"$/, '');
            return name;
          }
        }
      }
    } catch (error) {
      console.warn(`Warning: Could not parse app manifest for ${appId}: ${error}`);
    }

    return null;
  }

  /**
   * Get game name from library cache JSON (fallback 1).
   * 
   * @param appId - Steam app ID
   * @returns Game name or null if not found
   */
  private async getNameFromLibraryCache(appId: string): Promise<string | null> {
    const cacheFilePath = join(this.libraryCachePath, `${appId}.json`);
    
    if (!existsSync(cacheFilePath)) {
      return null;
    }

    try {
      const content = await readFile(cacheFilePath, 'utf-8');
      const data = JSON.parse(content) as LibraryCache;
      
      // Library cache has complex structure, look for descriptions
      for (const item of data) {
        if (Array.isArray(item) && item.length >= 2) {
          if (item[0] === 'descriptions' && typeof item[1] === 'object' && item[1] !== null) {
            const descData = (item[1] as any).data;
            if (descData && typeof descData === 'object') {
              // Try to extract name from description or other fields
              // This is a simplified approach - library cache is complex
              // TODO: Implement proper library cache parsing when needed
            }
          }
        }
      }
    } catch (error) {
      console.warn(`Warning: Could not parse library cache for ${appId}: ${error}`);
    }

    return null;
  }

  /**
   * Get game name from shortcuts.vdf (fallback 2).
   * 
   * @param appId - Steam app ID
   * @returns Game name or null if not found
   */
  private async getNameFromShortcuts(appId: string): Promise<string | null> {
    if (!existsSync(this.shortcutsPath)) {
      return null;
    }

    try {
      // TODO: Implement shortcuts.vdf parsing when VDF parser is available
      // For now, return null as this requires VDF binary parsing
      console.warn('Shortcuts.vdf parsing not yet implemented - requires VDF parser');
      return null;
    } catch (error) {
      console.warn(`Warning: Could not parse shortcuts.vdf: ${error}`);
    }

    return null;
  }

  /**
   * Get game name for an app ID using multiple fallback methods.
   * 
   * @param appId - Steam app ID
   * @param forceRefresh - Force refresh from sources, bypass cache
   * @returns Game name or fallback name
   */
  async getGameName(appId: string, forceRefresh: boolean = false): Promise<string> {
    // Check cache first (unless force refresh)
    if (!forceRefresh && this.cachedMappings.has(appId)) {
      return this.cachedMappings.get(appId)!;
    }

    // Try app manifest (primary method)
    let name = await this.getNameFromAppManifest(appId);
    if (name) {
      this.cachedMappings.set(appId, name);
      await this.saveCache();
      return name;
    }

    // Try library cache (fallback 1)
    name = await this.getNameFromLibraryCache(appId);
    if (name) {
      this.cachedMappings.set(appId, name);
      await this.saveCache();
      return name;
    }

    // Try shortcuts.vdf (fallback 2)
    name = await this.getNameFromShortcuts(appId);
    if (name) {
      this.cachedMappings.set(appId, name);
      await this.saveCache();
      return name;
    }

    // Try manual mappings (fallback 3)
    if (appId in this.manualMappings) {
      name = this.manualMappings[appId];
      this.cachedMappings.set(appId, name);
      await this.saveCache();
      return name;
    }

    // Final fallback
    const fallbackName = `App ${appId}`;
    this.cachedMappings.set(appId, fallbackName);
    await this.saveCache();
    return fallbackName;
  }

  /**
   * Check if an app is a system app (Proton, Steam Runtime, etc.).
   * 
   * @param appId - Steam app ID
   * @param appName - App name to check
   * @returns True if it's a system app
   */
  isSystemApp(appId: string, appName: string): boolean {
    const nameLower = appName.toLowerCase();
    return Array.from(this.systemApps).some(systemName => nameLower.includes(systemName));
  }

  /**
   * Refresh the entire cache by re-scanning all sources.
   * 
   * @returns Promise that resolves when cache is refreshed
   */
  async refreshCache(): Promise<void> {
    console.log('Refreshing app name cache...');
    this.cachedMappings.clear();
    await this.saveCache();
    console.log(`Cache refreshed. Found ${this.cachedMappings.size} mappings.`);
  }

  /**
   * Get the number of cached mappings.
   * 
   * @returns Number of cached mappings
   */
  getCacheSize(): number {
    return this.cachedMappings.size;
  }

  /**
   * Check if an app ID is in the cache.
   * 
   * @param appId - Steam app ID
   * @returns True if app ID is cached
   */
  isCached(appId: string): boolean {
    return this.cachedMappings.has(appId);
  }

  /**
   * Get all cached app IDs.
   * 
   * @returns Array of cached app IDs
   */
  getCachedAppIds(): string[] {
    return Array.from(this.cachedMappings.keys());
  }

  /**
   * Add a manual mapping for an app ID.
   * 
   * @param appId - Steam app ID
   * @param name - Game name
   */
  addManualMapping(appId: string, name: string): void {
    this.manualMappings[appId] = name;
    this.cachedMappings.set(appId, name);
  }

  /**
   * Remove a manual mapping for an app ID.
   * 
   * @param appId - Steam app ID
   */
  removeManualMapping(appId: string): void {
    delete this.manualMappings[appId];
    this.cachedMappings.delete(appId);
  }

  /**
   * Get the cache file path.
   * 
   * @returns Path to cache file
   */
  getCacheFilePath(): string {
    return this.cacheFile;
  }

  /**
   * Get the Steam user ID.
   * 
   * @returns Steam user ID
   */
  getSteamUserId(): string {
    return this.steamUserId;
  }

  /**
   * Get the Steam path.
   * 
   * @returns Steam installation path
   */
  getSteamPath(): string {
    return this.steamPath;
  }
}

File: backend/src/core/game_resolver.py
"""
Game name resolver for Steam app IDs.

This module handles resolving Steam app IDs to human-readable game names using multiple fallback methods.
"""

import json
import vdf
from pathlib import Path
from typing import Dict, Optional


class GameNameResolver:
    """Resolves Steam app IDs to human-readable game names using multiple fallback methods."""
    
    def __init__(self, steam_user_id: str = "107256425"):
        """
        Initialize the Game Name Resolver.
        
        Args:
            steam_user_id: Steam user ID (default: 107256425)
        """
        self.steam_user_id = steam_user_id
        self.steam_path = Path.home() / ".local" / "share" / "Steam"
        self.steamapps_path = self.steam_path / "steamapps"
        self.library_cache_path = self.steam_path / "userdata" / steam_user_id / "config" / "librarycache"
        self.shortcuts_path = self.steam_path / "userdata" / steam_user_id / "config" / "shortcuts.vdf"
        self.cache_file = Path("instance/data/app_name_cache.json")
        
        # Manual mappings for games not found in other sources
        self.manual_mappings = {
            "3241660": "R.E.P.O.",
            # Add more mappings as needed
        }
        
        # System apps to filter out (optional)
        self.system_apps = {
            "steam linux runtime", "proton", "steamworks", "steam client"
        }
        
        # Load cached mappings
        self.cached_mappings = self._load_cache()
    
    def _load_cache(self) -> Dict[str, str]:
        """Load cached app ID to name mappings."""
        if self.cache_file.exists():
            try:
                with open(self.cache_file, 'r', encoding='utf-8') as f:
                    return json.load(f)
            except Exception as e:
                print(f"Warning: Could not load app name cache: {e}")
        return {}
    
    def _save_cache(self) -> None:
        """Save app ID to name mappings to cache."""
        self.cache_file.parent.mkdir(parents=True, exist_ok=True)
        try:
            with open(self.cache_file, 'w', encoding='utf-8') as f:
                json.dump(self.cached_mappings, f, indent=2, ensure_ascii=False)
        except Exception as e:
            print(f"Warning: Could not save app name cache: {e}")
    
    def _get_name_from_app_manifest(self, app_id: str) -> Optional[str]:
        """Get game name from app manifest file (primary method)."""
        manifest_path = self.steamapps_path / f"appmanifest_{app_id}.acf"
        if not manifest_path.exists():
            return None
        
        try:
            with open(manifest_path, 'r', encoding='utf-8') as f:
                content = f.read()
                # Simple parsing for the name field
                for line in content.split('\n'):
                    if '"name"' in line and '\t' in line:
                        # Extract name from line like: "name"		"Game Name"
                        parts = line.split('\t')
                        if len(parts) >= 3:
                            name = parts[-1].strip().strip('"')
                            return name
        except Exception as e:
            print(f"Warning: Could not parse app manifest for {app_id}: {e}")
        
        return None
    
    def _get_name_from_library_cache(self, app_id: str) -> Optional[str]:
        """Get game name from library cache JSON (fallback 1)."""
        cache_file = self.library_cache_path / f"{app_id}.json"
        if not cache_file.exists():
            return None
        
        try:
            with open(cache_file, 'r', encoding='utf-8') as f:
                data = json.load(f)
                # Library cache has complex structure, look for descriptions
                for item in data:
                    if isinstance(item, list) and len(item) >= 2:
                        if item[0] == "descriptions" and isinstance(item[1], dict):
                            desc_data = item[1].get("data", {})
                            # Try to extract name from description or other fields
                            # This is a simplified approach - library cache is complex
                            pass
        except Exception as e:
            print(f"Warning: Could not parse library cache for {app_id}: {e}")
        
        return None
    
    def _get_name_from_shortcuts(self, app_id: str) -> Optional[str]:
        """Get game name from shortcuts.vdf (fallback 2)."""
        if not self.shortcuts_path.exists():
            return None
        
        try:
            with open(self.shortcuts_path, 'rb') as f:
                shortcuts_data = vdf.binary_load(f)
            
            # Look for matching shortcut by comparing with localconfig data
            # This is complex since we need to match by launch options or other criteria
            for shortcut_id, shortcut_info in shortcuts_data.get("shortcuts", {}).items():
                if shortcut_info.get("AppName"):
                    # For now, we'll use this as a general fallback
                    # In practice, we'd need to match by launch options or other criteria
                    pass
        except Exception as e:
            print(f"Warning: Could not parse shortcuts.vdf: {e}")
        
        return None
    
    def get_game_name(self, app_id: str, force_refresh: bool = False) -> str:
        """
        Get game name for an app ID using multiple fallback methods.
        
        Args:
            app_id: Steam app ID
            force_refresh: Force refresh from sources, bypass cache
            
        Returns:
            Game name or fallback name
        """
        # Check cache first (unless force refresh)
        if not force_refresh and app_id in self.cached_mappings:
            return self.cached_mappings[app_id]
        
        # Try app manifest (primary method)
        name = self._get_name_from_app_manifest(app_id)
        if name:
            self.cached_mappings[app_id] = name
            return name
        
        # Try library cache (fallback 1)
        name = self._get_name_from_library_cache(app_id)
        if name:
            self.cached_mappings[app_id] = name
            return name
        
        # Try shortcuts.vdf (fallback 2)
        name = self._get_name_from_shortcuts(app_id)
        if name:
            self.cached_mappings[app_id] = name
            return name
        
        # Try manual mappings (fallback 3)
        if app_id in self.manual_mappings:
            name = self.manual_mappings[app_id]
            self.cached_mappings[app_id] = name
            return name
        
        # Final fallback
        fallback_name = f"App {app_id}"
        self.cached_mappings[app_id] = fallback_name
        return fallback_name
    
    def is_system_app(self, app_id: str, app_name: str) -> bool:
        """Check if an app is a system app (Proton, Steam Runtime, etc.)."""
        name_lower = app_name.lower()
        return any(system_name in name_lower for system_name in self.system_apps)
    
    def refresh_cache(self) -> None:
        """Refresh the entire cache by re-scanning all sources."""
        print("Refreshing app name cache...")
        self.cached_mappings.clear()
        self._save_cache()
        print(f"Cache refreshed. Found {len(self.cached_mappings)} mappings.")
